src/state.rs

```diff
@@ -123,10 +163,53 @@ impl GameState {
         }
     }
 
-    pub fn update(&mut self, dt: f64) {
+    pub fn update(&mut self, world: &WorldState, dt: f64) {
+        let mut bullets = mem::replace(&mut self.bullets, vec![]);
+        'bullet: for bullet in bullets.iter_mut() {
+            bullet.update(dt);
+            // still inside world?
+            if bullet.position[0] > world.x || bullet.position[1] > world.y ||
+                bullet.position[0] < 0.0 || bullet.position[1] < 0.0 {
+                    continue;
+            }
+
+            for player in self.players.iter_mut() {
+                for unit in player.units.iter_mut() {
+                    if unit.health > 0 && unit.is_hit(UNIT_SIZE, bullet.position) {
+                        if unit.health > 10000 {
+                            unit.health -= 10000;
+                        } else {
+                            unit.health = 0;
+                        }
+                        println!("hit: {}", unit.health);
+                        continue 'bullet;
+                    }
+                }
+            }
+            self.bullets.push(bullet.clone());
+        }
         for player in self.players.iter_mut() {
-            for unit in player.units.iter_mut() {
-                unit.update(dt);
+            // remove units and add them back if still alive to avoid changing the vector while
+            // iterating
+            let mut units = mem::replace(&mut player.units, vec![]);
+            for unit in units.iter_mut() {
+                if unit.health > 0 {
+                    unit.update(dt);
+                    player.units.push(unit.clone());
+                } else {
+                    println!("killed unit!");
+                }
+            }
```

Comment:
This could be done cleanly with an iterator and filter. Maybe also for the bullets?

Reply:
Done.